### PR TITLE
[IE CLDNN] Fully Connected layer 3d support

### PR DESCRIPTION
**Pull request by ldebski**
Support for 3d format in Fully Connected layer in ref, mmad and tiled kernels.

JIRA: CVS-34199

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl [hunk 1]
```diff
@@ -279,14 +290,14 @@ KERNEL(fc)(
         for (uint bi = 0; bi < TILE_B; ++bi) {
             for (uint fi = 0; fi < TILE_OFM; ++fi) {
                 const bool should_write =
-                    OUTPUT_FEATURE_NUM %  (TILE_OFM * SIMD) == 0 ||
-                    out_f + fi * SIMD + get_sub_group_local_id() < OUTPUT_FEATURE_NUM;
+                    TILE_OUT_F_NUM %  (TILE_OFM * SIMD) == 0 ||
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl:293]: Please delete extra space after `TILE_OUT_F_NUM %`.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bfyx_ref.cl [hunk 2]
```diff
@@ -27,6 +27,27 @@ KERNEL(fc)(
 #endif
     )
 {
+#if OUTPUT_3D
+    const uint oxfm = get_global_id(0);
+    const uint b = get_global_id(1);
+    const uint oym = oxfm % OUTPUT_SIZE_Y;
+    const uint ofm = oxfm / OUTPUT_SIZE_Y;
+
+    ACCUMULATOR_TYPE dotProd = (ACCUMULATOR_TYPE)0;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bfyx_ref.cl:36]: I suppose you could use `ACCUMULATOR_TYPE dotProd = ACCUMULATOR_VAL_ZERO;` here.

File: inference-engine/thirdparty/clDNN/tests/test_utils/network_test.h [hunk 3]
```diff
@@ -300,6 +329,28 @@ class network_test {
         return add_node(id, reference_tensor_typed<T, 2>(output_data), { input, weights, bias });
     }
 
+    template <typename T, typename InputT, size_t InputN, typename WeightsT, typename BiasT>
+    typename reference_node<T, 4>::ptr add_fully_connected_3d(cldnn::primitive_id id,
+                                                           std::shared_ptr<reference_node<InputT, InputN>> input,
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/tests/test_utils/network_test.h:334]: Alignment

**Review by lznamens — CHANGES_REQUESTED** [review]
Also please update the license info for files `fully_connected.hpp` and `fully_connected_gpu_bfyx_ref.cl`.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp [hunk 4]
```diff
@@ -252,21 +282,42 @@ JitConstants FullyConnected_bf_tiled::GetJitConstants(const fully_connected_para
 
     jit.Merge(MakeConstantLoopUnrollJitConstants(dispatchData.tile_m));
 
-    bool realign_fp16_offset = params.inputs[0].GetDType() == Datatype::F16 && params.output.GetFirstElementOffset() % 2 != 0;
+    bool realign_fp16_offset = params.inputs[0].GetDType() == Datatype::F16 && params.inputs[0].GetFirstElementOffset() % 2 != 0;
+    // bool realign_fp16_offset = params.inputs[0].GetDType() == Datatype::F16 && params.inputs[0].Y().v % 2 != 0;
+    // realign_fp16_offset = false;
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl [hunk 5]
```diff
@@ -242,12 +253,12 @@ KERNEL(fc)(
...
     // Write results
-    uint output_offset = out_f * OUTPUT_FEATURE_PITCH + out_b * OUTPUT_BATCH_PITCH + OUTPUT_OFFSET;
+    uint output_offset = out_f * TILE_OUT_F_PITCH + out_b * TILE_OUT_B_PITCH + OUTPUT_OFFSET;
 
-    if (USE_BLOCK_WRITE && (OUTPUT_FEATURE_NUM % (TILE_OFM * SIMD) == 0 || out_f + (TILE_OFM * SIMD) <= OUTPUT_FEATURE_NUM)) {
+    if (USE_BLOCK_WRITE && (TILE_OUT_F_NUM % (TILE_OFM * SIMD) == 0 || out_f + (TILE_OFM * SIMD) <= TILE_OUT_F_NUM)) {
         #define WRITE_OUTPUT(bi) do {                                       \
                 OUTPUT_BLOCK_WRITE(output, output_offset, result[bi]);      \
-                output_offset += OUTPUT_BATCH_PITCH;                        \
+                output_offset += TILE_OUT_B_PITCH;                        \
```
[the remark this answers is not in the document]
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl (hunk 1 above)
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bf_tiled.cl]: Fixed

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bfyx_ref.cl (hunk 2 above)
Comment (ldebski) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fully_connected_gpu_bfyx_ref.cl]: Done

**Review by lznamens — APPROVED** [review]

**Pull request merged**